### PR TITLE
History: support classes modifying instances in __setstate__ 

### PR DESCRIPTION
**Pull request by perrinjerome**
Fixes the problem described on https://bugs.launchpad.net/zope2/+bug/735999 that still happens when comparing history of `Products.PythonScripts.PythonScript` created with an old script magic or python magic.

### CODE REVIEW

File: CHANGES.rst
Comment (dataflake) [inline CHANGES.rst:14]:
```suggestion

- Fix history page for classes modifying instances in ``__setstate__``,
```

**Review by dataflake — APPROVED** [review]
If you leave out the change in `History.py` and the tests fail with the same AssertionError explained in the Launchpad issue then that seems fine. I don't have much knowledge about the History stuff, though.

**Comment by perrinjerome** [comment]
> If you leave out the change in History.py and the tests fail with the same AssertionError explained in the Launchpad issue then that seems fine

I think so, on the launchpad issue, the description of the error was:

> ... it fails with an AssertionError in ZOPE/Connection.py:register on "assert obj._p_jar is self".

and without the change in `History.py` test errors like this:

```console
$ ./bin/test --test OFS.tests.testHistory
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.


Failure in test test_manage_historicalComparison (OFS.tests.testHistory.HistoryWithSetStateTest)
Traceback (most recent call last):
  File "/opt/slapos-shared/python3/6f204393301d3c4787363493d65ac803/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/slapos-shared/python3/6f204393301d3c4787363493d65ac803/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/srv/slapgrid/slappart3/srv/project/Zope/src/OFS/tests/testHistory.py", line 109, in test_manage_historicalComparison
    keys=[r[1]['key'], r[2]['key']]))
  File "/srv/slapgrid/slappart3/srv/project/Zope/src/OFS/History.py", line 192, in manage_historicalComparison
    rev1 = historicalRevision(self, serial)
  File "/srv/slapgrid/slappart3/srv/project/Zope/src/OFS/History.py", line 68, in historicalRevision
    rev.__setstate__(state)
  File "/srv/slapgrid/slappart3/srv/project/Zope/src/OFS/tests/testHistory.py", line 153, in __setstate__
    self.changed_something = True
  File "/srv/slapgrid/slappart3/srv/project/Zope/eggs/ZODB-5.8.0-py3.7.egg/ZODB/Connection.py", line 817, in register
    assert obj._p_jar is self
AssertionError

  Ran 9 tests with 1 failures, 0 errors and 0 skipped in 0.452 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
```

File: CHANGES.rst
Comment (perrinjerome) [reply CHANGES.rst:15]: Thanks, I applied this locally and push-forced

File: src/OFS/History.py
Comment (icemac) [inline src/OFS/History.py:59]: I think this can be left out: `register` was part of an old and no longer supported API, if I remember correctly.

**Review by icemac — APPROVED** [review]
I do not know much about the topic either but I have a little suggestion to try.

**Comment by icemac** [comment]
Additionally We need a rebase onto master when #1083 is merged to get MacOS tests fixed.

File: src/OFS/History.py
Comment (icemac) [reply src/OFS/History.py:59]: After reading the whole PR thread, I retract my comment. My memories seem to betray me. Time to get some fresh air.

**Comment by dataflake** [comment]
I just merged master with the button here. All green, this is good to go.

**Comment by perrinjerome** [comment]
> I just merged master with the button here. All green, this is good to go.

Thanks, so this means I should click merge. I will do in a few minutes /hours if no objectsion

**Comment by dataflake** [comment]
When tests are green and there are no unresolved conversations or change requests and you have approvals then you can merge at any time.

**Comment by d-maurer** [comment]
Michael Howitz wrote at 2022-12-23 00:52 -0800:
>@icemac approved this pull request.
>
>I do not know much about the topic either but I have a little suggestion to try.
>
>> @@ -56,7 +56,7 @@ def commit(self, object, transaction):
>     def abort(*args, **kw):
>         pass
>
>-    tpc_begin = tpc_finish = abort
>+    register = tpc_begin = tpc_finish = abort

`register` is still a `ZODB.Connection.Connection` method
(like the other methods).

**Comment by perrinjerome** [comment]
Thanks, I have been checking a bit more and I was not comfortable that this `raise` was not covered by the test:

https://github.com/zopefoundation/Zope/blob/a58333605a88ad9a24a7f07a820f4d837e2f7cd0/src/OFS/History.py#L52-L54

This code path has been unreachable since 2004 ( https://github.com/zopefoundation/ZODB/commit/61db515972d7a809ba2bb3588329a9b489890d0e ), I made more changes to cover this and keep the original intention of this `raise`: that attempting to commit a transaction changing old revisions of objects is an error.

To achieve this, I had to change the logic a bit. In that first version of the patch, modifying an old revision was just ignored because the object was not registered to the transaction, so this was a bit against the original idea of `TemporalParadox`.

In that new version, it is still possible to modify, but committing the transaction afterwards raises a `TemporalParadox` error.

To achieve this, I had to change `historicalRevision` to assign the `_p_jar` later, now for classes modifying instances in `__setstate__`, `historicalRevision` first create an object not associated with a database, call `__setstate__` and only then sets the `_p_jar`, this way, modification in `__setstate__` are ignored. This is actually similar to the patch that was attached to the launchpad issue.

**Comment by dataflake** [comment]
I can only repeat myself: My knowledge of the history mechanism and the intricate details of the ZODB is not enough to really review this. If the tests fail without the code changes and succeed with them in place I am guessing it's OK.

**Review by dataflake — APPROVED** [review]

**Comment by perrinjerome** [comment]
I'm merging this now, I wanted to be back from vacation before merging. Thanks again for the review and happy new year :)

**Pull request merged**
